File: 46-permutations/permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        def backtrack(cur):
            if len(cur) == len(nums):
                res.append(cur[:])
                return

            for num in nums:
                if num not in seen:
                    seen.add(num)
                    cur.append(num)
                    backtrack(cur)
                    seen.remove(cur.pop())

        res = []
        seen = set()
        backtrack([])

        return res

File: 46-permutations/permutations.php
class Solution {

	/**
	* @param Integer[] $nums
	* @return Integer[][]
	*/
	function permute($nums) {
		$n = count($nums);
		$ans = [];
		$set = [];
		$this->backtrack([], $ans, $set, $n, $nums);
		return $ans;
	}

	function backtrack($cur, &$ans, $set, $n, $nums) {
		if (count($cur) === $n) {
			$ans[] = $cur;
		}

		foreach ($nums as $num) {
			if (!array_key_exists($num, $set)) {
				$cur[] = $num;
				$set[$num] = True;
				$this->backtrack($cur, $ans, $set, $n, $nums);
				unset($set[array_pop($cur)]);
			}
		}
	}
}